### PR TITLE
feat(ipc) avoid new workers attempting to replay evicted events

### LINKED ISSUE #93: [enhancement] strengthen the IPC module
After some discussions last week with mlcache users, it appears that several improvements could be made to the IPC module in order to make it more robust for intensive use-cases:

- [x] First of all, document the `timeout` argument in the `cache:update()` method.
- [ ] Initialize `self.idx` not in `new()`, but at the time of the first call to `poll()`. The reason behind this is that users creating their cache (and IPC module) in `init_by_lua` will not benefit from the fix implemented in #88. When a worker crashes and the master process spawns a new worker, the `init_by_lua` context is not called, and the IPC `new()` method is not called again. This causes respawned workers to replay all events (events which have probably been evicted from the shm since then).
- [x] Add a guard in case `self.idx >= shm_idx`
- [ ] Investigate a "max timeouts threshold" values: when `poll()` timed out N times in a row, maybe we should bump `self.idx` all the way to `shm_idx` (thus assuming that all events in between `self.idx` and `shm_idx` have been discarded). This allows resuming normal polling operations, but means that some events may be left unprocessed.

### PR DESCRIPTION
**Pull request by thibaultcha**
A follow-up to #88.

Described in #93 and reported a few months ago is the likely possibility that mlcache and ipc instances are created in the `init` phase. If a worker is to be started much later in the master process' lifetime, the newly forked ipc instance will have an `idx` attribute set to the shm_idx value at the time of `init` (likely 0). These workers will resume polling evicted events, and `poll()` will likely timeout indefinitely from there.

For the fix in #88 to work, the mlcache instance has to be instantiated during `init_worker` or later.

This patch proposes an approach which works for instances created in both `init` and `init_worker`: as soon as the ipc shm has started evicting items, we guarantee that future workers will resume polling at the current index, without having to call any method but `poll()`.

Fix #93

### CODE REVIEW

File: lib/resty/mlcache/ipc.lua [hunk 1]
```diff
@@ -113,11 +105,21 @@ function _M:broadcast(channel, data)
...
-    local ok, err = self.dict:set(idx, marshalled_event)
+    local ok, err, forcible = self.dict:set(idx, marshalled_event)
     if not ok then
         return nil, "failed to insert event in shm: " .. err
     end
 
+    if forcible then
+        -- take note that eviction has started
+        -- we repeat this flagging to avoid this key from ever being
+        -- evicted itself
```
Comment (thibaultcha) [inline lib/resty/mlcache/ipc.lua:116]: A better approach here would be if `shm:incr()` accepted `flags`, or maybe even an `shm:set_flags()` API. In the absence of a solution to rely on flags, we rely on an extraneous key, and on it never reaching the LRU tail.

File: lib/resty/mlcache/ipc.lua [hunk 2]
```diff
@@ -113,11 +105,21 @@ function _M:broadcast(channel, data)
...
+        local ok, err = self.dict:set(FORCIBLE_KEY, true)
```
Comment (thibaultcha) [inline lib/resty/mlcache/ipc.lua:117]: A better approach here would be if shm:incr() accepted flags, or maybe even an shm:set_flags() API. In the absence of a solution to rely on flags, we rely on an extraneous key, and on it never reaching the LRU tail.

**Review by juancampa — COMMENTED** [review]
Wouldn't it be enough to run this on the first call to `poll`?
```
if self.idx < 0 then
  self.idx = shm_idx
end
``` 
:thinking:

**Comment by thibaultcha** [comment]
If we do this, we'd introduce a regression and we would prevent the first workers from polling any early broadcasted event, see for example [this test](https://github.com/thibaultcha/lua-resty-mlcache/blob/cf299e9e0b2914eb7449a6f13ad9642173cf7d06/t/00-ipc.t#L157-L191). Since what we're worried about are missing (evicted) events which would cause streaks of timeouts, I should still be safe to preserve this behavior that if the event at index 1 is still present, then we can just replay them all fast enough.

**Comment by thibaultcha** [comment]
Does it still sound good to you @juancampa? Something you'd care testing maybe? I doubt it for such a minor change but just checking-in in case.

**Comment by juancampa** [comment]
What I hesitate a bit on is that the behavior is significantly different between workers that are spawned before the first eviction and workers that are started after an eviction. A simple example would be:

a worker that is started, after 10000 events but _before_ the eviction happened. Wouldn't that worker go through all 10000 events? Compare that to a worker that was started after 10000 events but _after_ an eviction. This one wouldn't catch up at all.

For mlcache purposes, do we need to replay old events at all. If a worker is new, it won't have the keys being deleted by the `broadcast` in L1 anyway.

**Comment by thibaultcha** [comment]
Well, like I said above, new/fresh workers need to run past events before the first `poll()` in the current implementation, and I did not want to break this atm. Keeping track of which workers are new or which ones are re-spawning could be done but I think this is already a worthy improvement without too complicated of a patch. Better have a worker replay past events even if it may timeout a few times along the way as it catches up, rather than never catch up at all.

If there isn't any other way, maybe can change that behavior in a major release so that all ipcs first `poll()` always skip to the current index, for mlcache purposes only.

**Pull request merged**